File: chatroom/client/lib/utils/models/chats.dart
class Chat {
  int _id;
  bool _isGroup;
  String _groupName;
  String _groupDescription;
  Message _message;
  List<Users> _users;

  Chat(
      {int id,
      bool isGroup,
      String groupName,
      String groupDescription,
      Message message,
      List<Users> users}) {
    this._id = id;
    this._isGroup = isGroup;
    this._groupName = groupName;
    this._groupDescription = groupDescription;
    this._message = message;
    this._users = users;
  }

  int get id => _id;
  set id(int id) => _id = id;
  bool get isGroup => _isGroup;
  set isGroup(bool isGroup) => _isGroup = isGroup;
  String get groupName => _groupName;
  set groupName(String groupName) => _groupName = groupName;
  String get groupDescription => _groupDescription;
  set groupDescription(String groupDescription) =>
      _groupDescription = groupDescription;
  Message get message => _message;
  set message(Message message) => _message = message;
  List<Users> get users => _users;
  set users(List<Users> users) => _users = users;

  Chat.fromJson(Map<String, dynamic> json) {
    _id = json['id'];
    _isGroup = json['isGroup'];
    _groupName = json['groupName'];
    _groupDescription = json['groupDescription'];
    _message =
        json['message'] != null ? new Message.fromJson(json['message']) : null;
    if (json['users'] != null) {
      _users = new List<Users>();
      json['users'].forEach((v) {
        _users.add(new Users.fromJson(v));
      });
    }
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['id'] = this._id;
    data['isGroup'] = this._isGroup;
    data['groupName'] = this._groupName;
    data['groupDescription'] = this._groupDescription;
    if (this._message != null) {
      data['message'] = this._message.toJson();
    }
    if (this._users != null) {
      data['users'] = this._users.map((v) => v.toJson()).toList();
    }
    return data;
  }
}

class Message {
  int _id;
  String _date;
  String _text;
  int _userId;
  Sender _sender;

  Message({int id, String date, String text, int userId, Sender sender}) {
    this._id = id;
    this._date = date;
    this._text = text;
    this._userId = userId;
    this._sender = sender;
  }

  int get id => _id;
  set id(int id) => _id = id;
  String get date => _date;
  set date(String date) => _date = date;
  String get text => _text;
  set text(String text) => _text = text;
  int get userId => _userId;
  set userId(int userId) => _userId = userId;
  Sender get sender => _sender;
  set sender(Sender sender) => _sender = sender;

  Message.fromJson(Map<String, dynamic> json) {
    _id = json['id'];
    _date = json['date'];
    _text = json['text'];
    _userId = json['userId'];
    _sender =
        json['sender'] != null ? new Sender.fromJson(json['sender']) : null;
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['id'] = this._id;
    data['date'] = this._date;
    data['text'] = this._text;
    data['userId'] = this._userId;
    if (this._sender != null) {
      data['sender'] = this._sender.toJson();
    }
    return data;
  }
}

class Sender {
  String _name;
  String _surname;
  int _id;
  String _nickname;
  int _imageId;

  Sender({String name, String surname, int id, String nickname, int imageId}) {
    this._name = name;
    this._surname = surname;
    this._id = id;
    this._nickname = nickname;
    this._imageId = imageId;
  }

  String get name => _name;
  set name(String name) => _name = name;
  String get surname => _surname;
  set surname(String surname) => _surname = surname;
  int get id => _id;
  set id(int id) => _id = id;
  String get nickname => _nickname;
  set nickname(String nickname) => _nickname = nickname;
  int get imageId => _imageId;
  set imageId(int imageId) => _imageId = imageId;

  Sender.fromJson(Map<String, dynamic> json) {
    _name = (json['name']).toString().split(" ").map((str) => str.substring(0, 1).toUpperCase() + str.substring(1)).join(" ");
    _surname = (json['surname']).toString().split(" ").map((str) => str.substring(0, 1).toUpperCase() + str.substring(1)).join(" ");
    _id = json['id'];
    _nickname = json['nickname'];
    _imageId = json['imageId'];
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['name'] = this._name;
    data['surname'] = this._surname;
    data['id'] = this._id;
    data['nickname'] = this._nickname;
    data['imageId'] = this._imageId;
    return data;
  }
}

class Users {
  int _id;
  int _imageId;
  String _name;
  String _surname;

  Users({int id, int imageId, String name, String surname}) {
    this._id = id;
    this._imageId = imageId;
    this._name = name;
    this._surname = surname;
  }

  int get id => _id;
  set id(int id) => _id = id;
  int get imageId => _imageId;
  set imageId(int imageId) => _imageId = imageId;
  String get name => _name;
  set name(String name) => _name = name;
  String get surname => _surname;
  set surname(String surname) => _surname = surname;

  Users.fromJson(Map<String, dynamic> json) {
    _id = json['id'];
    _imageId = json['imageId'];
    _name = (json['name']).toString().split(" ").map((str) => str.substring(0, 1).toUpperCase() + str.substring(1)).join(" ");
    _surname = (json['surname']).toString().split(" ").map((str) => str.substring(0, 1).toUpperCase() + str.substring(1)).join(" ");
  }

  Map<String, dynamic> toJson() {
    final Map<String, dynamic> data = new Map<String, dynamic>();
    data['id'] = this._id;
    data['imageId'] = this._imageId;
    data['name'] = this._name;
    data['surname'] = this._surname;
    return data;
  }
}

File: chatroom/server/modules/socket/data/chats.js
const logger = require('../../logger');
const database = require('../../database');

/**
 * Moduli utilizzati per la gestione delle chats
 */
module.exports = {
  get: async (socket, data) => {
    if (true) { //TODO: verificare se il client ha inviato un TOKEN valido
      let userId = database.escape(data.id);
      let chatsDb = await database.query("SELECT chats.isGroup, chats.groupDescription, chats.groupName, recipients.userId, recipients.chatId, recipients.name AS userName, recipients.surname AS userSurname, recipients.nickname AS userNickname, recipients.lastAccessDate AS userLastAccessDate, recipients.imageId AS userImageId FROM	users INNER JOIN userChatAssociations ON users.id = userChatAssociations.userId INNER JOIN chats ON userChatAssociations.chatId = chats.id INNER JOIN (SELECT userChatAssociations.userId, userChatAssociations.chatId, users.name, users.surname, users.nickname, users.lastAccessDate, users.imageId FROM userChatAssociations INNER JOIN (SELECT userChatAssociations.chatId FROM userChatAssociations WHERE userChatAssociations.userId = " + userId + ") chatConUno ON chatConUno.chatId = userChatAssociations.chatId INNER JOIN users ON userChatAssociations.userId = users.id WHERE userChatAssociations.userId != " + userId + ") recipients ON recipients.chatId = userChatAssociations.chatId WHERE users.id = " + userId);
      if (chatsDb.length > 0) {
        let chats = [];

        for (let i = 0; i < chatsDb.length; i++) {
          //TODO: fare una join per il todo sotto
          let message = await database.query("SELECT messages.id, messages.text, messages.date, messages.chatId, messages.userId FROM messages INNER JOIN chats ON messages.chatId = chats.id WHERE chats.id = " + chatsDb[i].chatId + " ORDER BY messages.date DESC LIMIT 1");
          if (chatsDb[i].isGroup == 1) { //chatroom
            if (!contains(chats, chatsDb[i].chatId)) {
              chats.push({
                id: chatsDb[i].chatId,
                isGroup: true,
                groupName: chatsDb[i].groupName,
                groupDescription: chatsDb[i].groupDescription,
                message: {
                  id: message[0].id,
                  date: message[0].date,
                  text: message[0].text,
                  userId: message[0].userId,
                  sender: { //TODO: i dati non sono corretti
                    name: chatsDb[i].userName,
                    surname: chatsDb[i].userSurname,
                    id: chatsDb[i].userId,
                    nickname: chatsDb[i].userNickname,
                    imageId: chatsDb[i].userImageId
                  }
                },
                users: await database.query("SELECT users.id, users.imageId, users.name, users.surname, users.nickname FROM users INNER JOIN userChatAssociations ON userChatAssociations.userId = users.id WHERE userChatAssociations.chatId=" + chatsDb[i].chatId + " AND users.id != " + userId),
              });
            }
          } else { //chat with 2 users
            chats.push({
              id: chatsDb[i].chatId,
              isGroup: false,
              groupName: null,
              groupDescription: null,
              message: {
                id: message[0].id,
                date: message[0].date,
                text: message[0].text,
                userId: message[0].userId,
                sender: { //TODO: i dati non sono corretti
                  name: chatsDb[i].userName,
                  surname: chatsDb[i].userSurname,
                  id: chatsDb[i].userId,
                  nickname: chatsDb[i].userNickname,
                  imageId: chatsDb[i].userImageId
                }
              },
              users: [{
                id: chatsDb[i].userId,
                imageId: chatsDb[i].userImageId,
                name: chatsDb[i].userName,
                surname: chatsDb[i].userSurname
              }],
            });
          }
        }

        chats.sort((x, y) => (new Date(x.message.date).getTime() > new Date(y.message.date).getTime()) ? -1 : 1);
        socket.write(JSON.stringify({
          status: 200,
          event: 'chats',
          data: chats
        }));

        socket.event = "chats";
        socket.token = data.token;
        require('../../../app')._CONNECTIONS.add(socket);

      } else {
        socket.write(JSON.stringify({ event: 'chats', status: 204, error: 'Non hai ancora nessuna chat! Inizia a scrivere a qualcuno!' }));
      }
    } else {
      socket.write(JSON.stringify({ event: 'chats', status: 401, error: 'Non sei autorizzato a visionare queste chat!' }));
    }
  }
}

/**
 * Funzione utilizzata per verificare se è già presente una chat con un determinato 
 * id
 * 
 * @param {Array} chats 
 * @param {int} chatId 
 */
function contains(chats, chatId) {
  for (let i = 0; i < chats.length; i++) {
    if (chats[i].id == chatId) {
      return true;
    }
  }

  return false;
}